Sources/NIO/Embedded.swift

```diff
@@ -325,7 +320,7 @@ public class EmbeddedChannel: Channel {
     // Embedded channels never have parents.
     public let parent: Channel? = nil
 
-    public func readOutbound() -> IOData? {
+    public func readOutbound() -> NIOAny? {
```

Comment:
Returning `NIOAny` from here is really painful, because without importing NIO as `@testable` you cannot unwrap this to the appropriate type (the various `tryAs` functions are internal).

This should use the same pattern as `readInbound`, which is that it should be generic and return a `T`.

Reply:
That said, it may be helpful to adopt a change we've made elsewhere, which is that this and `readInbound` could take an optional type parameter. That is:

```swift
public func readOutbound<T>(as type: T.Type) -> T?
    return self.readFromBuffer(buffer: &self.channelCore.outboundBuffer)
}
```

The same change could be made to `readInbound` as well, which will help clear up the testing changes.

Reply:
Great solution tnx.
Did you mean `public func readOutbound<T>(as type: T.Type = T.self) -> T?`?

Reply:
Also in the meantime I ran into an issue. You would not be able to ever use
`XCTAssertNotNil(channel.readOutbound())`
since that would try to get `Any` which would crash when trying to read `.other` while you actually have say a `ByteBuffer` stored. How should this be addressed(if it should be)?

Reply:
@JovanMilenkovic yes, `public func readOutbound<T>(as type: T.Type = T.self) -> T?` is perfect.

So maybe we could make the `tryAsAny` case a bit smarter and have it no mandate the `.other` case? I think if we request `Any` but find `.byteBuffer` we should still handle this properly. But I'm happy if we don't do this in this PR but in a separate one instead.

Reply:
I’m not sure at what point is semantics broken. If i `tryAsOther` and request a `ByteBuffer` should it work? The  case and `tryAsOther`

Reply:
I’m not sure at what point is semantics broken. If i `tryAsOther` and request a `ByteBuffer` should it work? 
The storage cases and `tryAsOther` are at the same access control level. Any input?

Reply:
> If yes, how does it help with clearing testing changes since there are no changes to `readInbound` methods?

You're right, we'll still need to change those. However, they'll be closer to the style of `readInbound` than this proposed change.

> Also would it be preferred to use the method parameter unless the type is already defines somewhere above?

Yes.

> How should this be addressed(if it should be)?

You write this:

```swift
XCTAssertNotNil(channel.readOutbound(as: IOData.self))
```

Or whatever is the correct type. You are never allowed to not say what type you want. This is a feature, not a bug, IMO.

> The storage cases and `tryAsOther` are at the same access control level. Any input?

I think ignore the `tryAs` functions: they are not for NIO users, they are specifically used internally to NIO. We should not expect users to call them.

Reply:
The thing that looks like a bug is that you say that you want `Any`(or the compiler infers it) and you get an assert if storage is not on `.other` and the message is quite cryptic with you tried to cast to `Any` but failed. That said, I’ve tried both and I think it should be explicit as you said, it doesn’t seem like it’s worth complicating `NIOAny`